docs/it/parametri_scope_claims.rst

```diff
@@ -23,13 +23,13 @@ Nel caso di utilizzo del parametro **scope** i seguenti valori sono supportati:
 Gli attributi richiesti tramite il parametro **scope** sono disponibili sia nell'ID Token e sia nella risposta allo userinfo endpoint.
 
 .. note::
-    Il parametro **scope** PUÒ contenere uno o più valori separati da uno spazio.
+    Il parametro **scope** PUÒ contenere uno o più valori separati da uno spazio. Ad esempio l'utilizzo congiunto di *profile* e *email* permette di ottenere l'unione degli insiemi degli attributi (Minimum Dataset eIDAS e l'email).
 
 Nel caso di richiesta di singoli attributi dell'utente o specifiche combinazioni di essi, Il RP PUÒ usare il parametro **claims**. 
 Per la definizione del parametro **claims** e la modalità di utilizzo per la richiesta degli attributi dell'utente si può fare riferimento a `OpenID.Core#ClaimsParameter`_. 
 
 .. warning::
-    - Nell'oggetto *id_token* del parametro **claims** è possibile richiedere solo il Minimum Dataset eIDAS. Gli altri attributi dell'utente DEVONO essere richiesti nell'oggetto *userinfo* del parametro **claims**.  
+    - Nell'oggetto *id_token* del parametro **claims** è possibile richiedere solo il Minimum Dataset eIDAS. Gli altri attributi dell'utente DEVONO essere richiesti nell'oggetto *userinfo* del parametro **claims**. Inoltre, gli attributi utente richiesti nell'oggetto *id_token* sono disponibili anche allo *userinfo endpoint*. 
```

Comment:
@agcolella in SPID OIDC guidelines, none user attributes are expected. It's not possible to request user attributes in the id_token, so neither the attributes related to the eIDAS Minimum Dataset

Reply:
We know that this is it to date, that we have a pending task in AgID to enable this mapping (see my email of two days ago) and that, to unblock this PR we may consider to move this section in a great warning "only for cie"

Let's wait until tomorrow meeting with the PMO to have an agenda and a decision for this

Reply:
@damikael AgID ha normato questo aspetto qui
https://www.agid.gov.it/sites/default/files/repository_files/tabella_attributi_v.1.3.pdf

possiamo procedere con l'approvazione di questa PR?

Reply:
@peppelinux nella tabella attributi è normata la corrispondenza dei claim e lo scope profile corrispondente all'eIDAS Minimum Dataset, ma non che gli attributi possano ritornare nell'id_token. 

Inoltre, in OIDC Core si legge: "_The Claims requested by the profile, email, address, and phone scope values are returned from the UserInfo Endpoint, as described in [Section 5.3.2](https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse), when a response_type value is used that results in an Access Token being issued._" 

e in OIDC SPID l'unico flusso consentito è AuthCodeFlow.

Per non creare confusione, secondo me è preferibile spostare in "only for cie". Per l'approvazione o meno lascio a @agcolella .

Reply:
Capisco, dobbiamo ritenere aperto ancora il task relativo alla abilitazione dei claim all'interno dell'id token, considerando che questo non esce fuori dal RP e non è usabile a scopi di interoperabilità verso sistemi terzi.

Questo task è stato richiesto il 12 Aprile 2022 e mi sembra essere l'unico aperto della lista che abbiamo ad oggi risolto.
@nunzionapoli @agcolella ^

Reply:
@damikael , in OIDC Core non trovo nessun testo normativo che dice che i claim devono essere disponibili esclusivamente nello userinfo. D'altro canto l'utilizzo degli scope per come è descritto mi pare compliant a OIDC Core in quanto gli attributi dell'utente sono disponibili nello userinfo (e ANCHE nell'ID token ovviamente). Non capisco come questo possa generare confusione. 
Anzi, proprio per evitare ogni confusione stiamo valutando di rendere disponibili tutti gli attributi utente richiesti sia nello userinfo e sia nell'ID token proprio in ottica di semplificazione.

Reply:
@fmarino-ipzs la norma è stabilita nelle LL.GG. OIDC SPID.
"Non è possibile richiedere attributi SPID nell’id_token" (Rif. LL.GG. OIDC SPID - 5.1)
Per confusione, intendo quella che potrebbe venirsi a creare leggendo da una parte che non è possibile richiedere attributi nell'id_token, e dall'altra che l'id_token può ritornare attributi.

Reply:
Certo capisco, ma nel precedente commento avevi fatto riferimento a OIDC Core e da questo punto di vista non vedo elementi normativi che non consentano l'utilizzo dell'ID Token per veicolare gli attributi utente. Le mie considerazioni, del tutto personali, sono le seguenti:
- L'adozione delle LL.GG è ancora in una fase iniziale;
- la modifica richiederebbe davvero pochissimo sforzo implementativo degli OP a fronte di una semplificazione lato integratori;
- in linea di principio strumenti normativi per effettuare modifiche o precisazioni alle LL.GG esistono (es. Avvisi) e sono stati già usati diffusamente anche per SAML da quanto mi risulta ( diverso sarebbe stato se quanto da noi proposto andasse contro OIDC Core);
- non vedo particolari problemi dal punto di vista della sicurezza.
Detto ciò ovviamente non ho problemi a creare dei box di disambiguazione SPID/CIE per questi aspetti.

Reply:
Per SPID dall'inizio si è pensato di veicolare gli attributi utente esclusivamente in userinfo, proprio poichè in tal modo possono essere cifrati. 
Avere gli attributi utente cifrati risulta utile nel caso degli aggregatori, in cui l'implementazione potrebbe essere centralizzata sull'aggregatore, ma il dato trattato solo dall'aggregato. 
In una tale implementazione il JWT cifrato ottenuto come response dalla userinfo può essere inviato all'aggregato in modo che gli attributi in esso contenuti possono essere decifrati esclusivamente dall'aggregato, con la propria chiave privata, restando invece nascosti all'aggregatore.
In questo caso, permettere gli attributi in id_token, non cifrato, significherebbe, invece, esporre le informazioni dell'utente all'aggregatore senza motivo.
Riguardo alla modifica delle LL.GG., gli avvisi consentono di integrare o precisare quanto indicato nelle LL.GG. ma non sono sicuro che una modifica sostanziale al testo (le LL.GG. indicano esplicitamente "Non è possibile richiedere attributi SPID nell’id_token") possa essere apportata con un semplice Avviso.

Reply:
A parte che avrei qualche perplessità sulla concreta utilità di questo caso d'uso (vedi SA full che hanno le chiavi private degli aggregatori). Ma in ogni caso possiamo prevedere di avere gli attributi solo nello userinfo utilizzando il parametro claims (per i SA light che vogliono avere una implementazione privacy preserving). Questo risolvete il problema da te posto garantendo agli altri integratori di avere un approccio semplificato tramite lo scope.

Reply:
Non so se mi sta sfuggendo qualcosa ma il token endpoint è protetto da private_key_jwt il che vuol dire che chi chiama l'endpoint deve avere la chiave privata. Due sono le possibilità che vedo:
1. La chiave privata del RP è gestita e posseduta dal SA: in questo caso il SA sarebbe in grado di ottenere gli attributi utente restituiti dallo userinfo e la sua cifratura non darebbe nessun valore aggiunto in termini di privacy;
2. La chiave privata è custodita solo dal RP aggregato: in questo caso solo RP può autenticarsi al token endpoint e dunque ottiene lui direttamente anche L'ID token. In questo caso il SA non ha accesso all'ID token e il tema privacy nemmeno si pone.
Sto sbagliando qualcosa io?

Reply:
@damikael il caso d'uso dei token criptati per gli aggregatori è stato superato con OIDC Federation 1.0, questo requisito di id_token svestiti di claim e criptati ricordo che fosse venuto di conseguenza meno.

Abbiamo un thread con Enrica sulla richiesta di avvicinare quanto più possibile SPID OIDC allo standard e come sai stiamo facendo il possibile anche con OIDF per portare avanti iGOV con i nostri contributi.

Questo aspetto è importante e vi chiedo con quale strumento normativo si possa risolvere e in che tempi @agcolella 
Abbiamo deciso di non gestirlo nel PM proprio perché a noi risulta alla stregua di un "dettaglio" facilmente superabile, come da ns ultimo meeting 

@nunzionapoli ^

Reply:
@fmarino-ipzs , le chiavi di firma e di cifratura possono essere, ed è raccomandato che lo siano, diverse. Ovvero nel jwks il relativo jwk riporta nel claim _use_ se si tratta di _sig_ o _enc_. Pertanto riguardo al tuo punto 1, l'aggregatore può autenticarsi al token endpoint firmando client_assertion con la sua chiave di firma, mentre il jwt ricevuto da userinfo sarebbe cifrato con la chiave di cifratura la cui chiave pubblica è esposta nel metadata dell'aggregato, ma la cui chiave privata è in possesso esclusivamente dell'aggregato.
Tale caso d'uso permette di avere tutta l'implementazione gestita dall'aggregatore, quindi l'aggregato non deve preoccuparsi di implementare alcuna chiamata e di conseguenza i relativi controlli (come sarebbe necessario nel caso relativo al tuo punto 2) ma al contempo il soggetto aggregatore non tratta dati eccedenti.

Reply:
@peppelinux in OIDC SPID id_token non è cifrato, ma solo userinfo è cifrato. Potresti indicare meglio come OIDC Fed risolverebbe il caso che ho esposto sopra, ovvero: implementazione gestita interamente dall'aggregato e dati dell'utente leggibili solo dall'aggregato.
Sono sempre d'accordo nel seguire lo standard (e in effetti quanto stabilito nelle LL.GG. OIDC non è in contraddizione con lo standard, infatti non è obbligatorio che i claim debbano essere presenti in id_token, se è previsto l'utilizzo di userinfo)
ma sono anche dell'idea che la sola semplificazione non è detto sia sempre la soluzione migliore.
In ogni caso, ritengo che una qualsiasi decisione deve comunque seguire un iter corretto e ordinato, pertanto, se si decide di modificare quanto normato nelle attuali LL.GG., prima di scrivere specifiche contrastanti in questa o altra documentazione, occorre prima rettificare le LL.GG. secondo le modalità previste dall'agenzia.

Reply:
@damikael, ammetto che è la prima volta che sento uno scenario di questo tipo che mi pare molto interessante anche se non so quanto lo sia nella pratica. in base alle implementazioni di SA che conosco io di solito forniscono ai RP tutto il layer IAM che comprende anche gli attributi utente con i quali si autentica. Probabilmente le mia conoscenze sono limitate ma in ogni caso si può ottenere lo stesso risultato pur senza rinunciare alla possibilità di avere gli attributi nell'ID token. Infatti nello scenario da te proposto è sufficiente non usare gli scope ma il parametro claims con gli attributi utente solo nel parametro userinfo. In questo modo nell'ID Token ci sarà solo il sub pairwise e tutto viaggia nello userinfo cifrato. Per tutti i casi dove non ci sono privacy concerns direi che avere la possibilità di ottenere gli attributi direttamente nell'ID token (oltre che nello userinfo) costituisce una utile semplificazione. Che ne pensate?

Reply:
@fmarino-ipzs , preciso che le mie sono solo considerazioni tecniche che emergono dal concetto generale, perseguito durante la definizione delle LL.GG., che gli attributi utente debbano ritornare solo crittografati in maniera tale che possano essere letti esclusivamente dal soggetto al quale sono destinati. Questo permette di ottenere privacy _by design_, indipendentemente dalle possibili soluzioni implementative.
La soluzione da te proposta è per me condivisibile ma farebbe venir meno questo concetto, in quanto, tornando al caso d'uso d'esempio, il soggetto aggregatore potrebbe, tramite lo scope, ottenere comunque la possibilità di "inserirsi" nella lettura degli attributi dell'utente, potenzialmente senza l'approvazione dell'aggregato.
A prescindere dalle mie personali considerazioni, per procedere ad una eventuale modifica delle LL.GG. in tal senso, proprio per il potenziale impatto sulla privacy, credo sia importante anche il parere del GPDP.

Reply:
un soggetto aggregatore full non usa ne possiede le chiavi private di un aggregato ma piuttosto usa le proprie chiavi per creare le richieste di autorizzazione e di client auth su token endpoint (private_key_jwt). Un aggregatore full può emettere le chiavi per il proprio aggregato senza approvvigionargliele. Un OP cripta il Nested JWT di userinfo usando la pubblica del soggetto aggregatore. L'aggregato riceve i claim dell'utente secondo uno schema di interoperabilità end to end con l'aggregatore full (che fa da proxy). Si faccia esempio di un aggregatore full che aggrega un SP SAML2, pubblicando l'entity configuration di questo.

Nel caso di aggregatore light il soggetto aggregatore è un intermediario di federazione puro, come da OIDC Fed.

Non c'è il requisito di criptare la userinfo se non per il fatto che il bearer token (access token) può essere rubato e la criptazione di questo JWT equivale ad un qualsiasi altro meccanismo di PoP.

id token non è un bearer e non viene utilizzato per scopi di interoperabilità, in altre parole non "esce fuori" dal RP.

per questa ragione pur criptando id token, come consentito da OIDC Core o iGov, questo verrebbe criptato comunque con la pubblica dell'aggregatore full.

Fatte queste assunzioni possiamo ritenere necessario abilitare i claim anche in id_token e di superare le attuali LG SPID perché i requisiti di criptazione di id_token o rimozione dei claim utente, alla luce delle precedenti assunzioni, decadono.

Riconosciamo quanto definito nelle LG SPID OIDC e al contempo riteniamo che due nuovi requisiti risultino di maggiore interesse. Lo scopo di questi è quello di migliorare l'adesione di SPID a OIDC Core e iGov e quello di conformare i profili implementativi di SPID e CIE. Questi sono requisiti di interesse strategico per una migliore uniformità di SPID agli standard internazionali e nazionali, allo scopo di ridurre le differenze e i costi implementativi che gli svilppatori devono gestire per adottare una implementazione sia per SPID che per CIE che per qualsiasi altra implementazione OIDC standard.

per adesso, per sbloccare questa PR, chiedo a @fmarino-ipzs di aggiungere un box di disambiguazione e ad Antonio Colella di indicarci i tempi e i modi con i quali gestire questa integrazione, mediante avviso o altro atto, all'interno profilo OIDC SPID. Se necessario possiamo gestirlo nella milestone corrente del PM relativo a OIDC SPID.

Reply:
Done here
https://github.com/italia/spid-cie-oidc-docs/pull/51/commits/0bc962d37e05b78c51eaa0206b901d7419151d06

a seguito dei riscontri con @agcolella tratteremo questo aspetto in SPID con calma e in seguito, gestendolo nel PM